### PR TITLE
redis_proxy: Fix #35983 - External Authentication should respect pipelining

### LINKED ISSUE #35983: redis_proxy: External Authentication should support pipelining
*Title*: Redis Proxy's External Authentication should support pipelining

*Related PR*
https://github.com/envoyproxy/envoy/pull/35643

*Description*:
We started using External Authentication and it works exactly as expected on a normal use-case. But on some setups, we notice that if an `AUTH` command is pipelined with another command, since the gRPC call is async, it may process another command before `AUTH` is fully processed, which by Redis standards should not be possible, ordering should be respected.

*Repro steps*:
1. Boot an Envoy Redis proxy with external authentication to a gRPC service
2. Send the following payload:
```
*2
$4
AUTH
$<pwd-size>
<valid-password>
*1
$4
PING
```
Expected result:
```
+OK
+PONG
```

Obtained result:
```
+OK
-NOAUTH Authentication required.
```

*Fix proposal*
I would be totally open to prepare a fix, since I'm familiar with the external auth code anyway. But I just would need some guidance on whether to:

1. Create a request queue until the result of AUTH. Basically we already have in the connection the state of the auth request (`Pending` or `Ready`). After the command splitter is executed (and it's not an AUTH command), if the state is pending, we put the command on a queue/buffer with configurable size limit.
2. Make the gRPC call synchronous (not good idea I think)

In the meantime I'll try to open a PoC PR for option 1.

Would love to have your input 😄 
cc @mattklein123 , @phlax

### PR DESCRIPTION
**Pull request by dbarbosapn**
**Description**
We started using External Authentication and it works exactly as expected on a normal use-case. But on some setups, we notice that if an AUTH command is pipelined with another command, since the gRPC call is async, it may process another command before AUTH is fully processed, which by Redis standards should not be possible, ordering should be respected.

_Risk Level_: Low (small change to unreleased feature)

_Testing_:

**Unit Tests:** ✅ 

**Integration Tests:** ✅

**Manual Testing (repro):** ✅ 
Before fix:
image

After fix:
image

Fixes #35983

### CODE REVIEW

**Comment by dbarbosapn** [comment]
Hi @mattklein123!
Follow-up on previous PR. Our internal testing identified a bug on command pipelining. We're using this build now, everything looks good 😄 
Would appreciate your review 🙇

File: source/extensions/filters/network/redis_proxy/proxy_filter.h [hunk 1]
```diff
@@ -139,6 +139,8 @@ class ProxyFilter : public Network::ReadFilter,
     Common::Redis::Client::Transaction& transaction() override { return parent_.transaction(); }
 
     ProxyFilter& parent_;
+    // This value is set when the request is on hold, waiting for an external auth response.
+    Common::Redis::RespValuePtr pending_value_;
```
Comment (mattklein123) [inline source/extensions/filters/network/redis_proxy/proxy_filter.h:143]: nit: I would call this `pending_request_value_` or just `pending_request_`. Technically I think you could merge `pending_value_` and `pending_response_` as I don't think they can ever be set at the same time, but not sure that is worth it.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
LGTM with small nit.

/wait

File: source/extensions/filters/network/redis_proxy/proxy_filter.h (hunk 1 above)
Comment (dbarbosapn) [reply source/extensions/filters/network/redis_proxy/proxy_filter.h:143]: The problem with merging is that there is a condition where if `pending_response_` is set, the pending responses get flushed, when `onResponse` is called, and the request is removed from queue.
Renaming to `pending_request_value_` as suggested 😄

**Comment by dbarbosapn** [comment]
> LGTM with small nit.
> 
> /wait

Addressed 😄 Thank you!

**Pull request merged**